sprint #13

Graev opened the pull request:
в .eslintrc добавлено исключение для _id;
Node.js приложение подключается к серверу Mongo по адресу mongodb://localhost:27017/mestodb;
создана схема и модель пользователя с полями name, about и avatar, поля корректно валидируются;
создана схема и модель карточки с полями name, link, owner, likes и createdAt, поля корректно валидируются;
запрос на GET /users возвращает всех пользователей из базы;
запрос GET /users/:userId возвращает конкретного пользователя;
запрос POST /users создаёт пользователя;
запрос GET /cards возвращает все карточки всех пользователей;
запрос POST /cards создаёт карточку;
запрос PATCH /users/me обновляет профиль;
запрос PATCH /users/me/avatar обновляет аватар;
запрос PUT /cards/:cardId/likes — поставит лайк карточке;
запрос DELETE /cards/:cardId/likes — уберет лайк карточке;
если в любом из перечисленных запросов что-то идёт не так, сервер возвращает ошибку, не забывайте добавлять статус;
папки public и data удалены;

Graev committed:
add project to GH

Graev committed:
Error correction

Справил метод подключение JSON файлов.
Изменил метод поиска пользователя по id.

Graev committed:
sprint #13

в .eslintrc добавлено исключение для _id;
Node.js приложение подключается к серверу Mongo по адресу mongodb://localhost:27017/mestodb;
создана схема и модель пользователя с полями name, about и avatar, поля корректно валидируются;
создана схема и модель карточки с полями name, link, owner, likes и createdAt, поля корректно валидируются;
запрос на GET /users возвращает всех пользователей из базы;
запрос GET /users/:userId возвращает конкретного пользователя;
запрос POST /users создаёт пользователя;
запрос GET /cards возвращает все карточки всех пользователей;
запрос POST /cards создаёт карточку;
запрос PATCH /users/me обновляет профиль;
запрос PATCH /users/me/avatar обновляет аватар;
запрос PUT /cards/:cardId/likes — поставит лайк карточке;
запрос DELETE /cards/:cardId/likes — уберет лайк карточке;
если в любом из перечисленных запросов что-то идёт не так, сервер возвращает ошибку, не забывайте добавлять статус;
папки public и data удалены;

Graev committed:
Added catching of the lack of user/card

Добавил отлов отсутствие пользователя/карточки

Graev committed:
Correct the error

Boortcore commented:
Необходимо исправить:
- в файлах проекта присутствуют мелкие ошибки линтинга. Например,
![image](https://user-images.githubusercontent.com/17438227/69229426-725e3d00-0b96-11ea-8061-c2368833cd07.png)
- API возвращает 200 при попытке получить юзера по валидному ObjectId "5dd510024c38a003fc91216b", которого не существует в базе 
- API создаёт карточку без свойства link
- API возвращает 200 при попытке удалить карточку по валидному ObjectId "5dd4e2b98151d0076cc9ae5a", которого не существует в базе
- при попытке удалить карточку по невалидному id сервер выдаёт предупреждение: " UnhandledPromiseRejectionWarning: Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client"

Можно лучше:
- в папке `routes` можно создать файл index.js, объявить там роутер и через метод `use` подключить роуты cards и users к этому роуту. После чего экспортировать только один роутер из папки и подключить его в app.js. Это сделает корневой файл немного чище.
- При создании карточки или юзера правильнее возвращать код ответа `201`.
- лучше использовать промисы, а не колбэки. 
![image](https://user-images.githubusercontent.com/17438227/69230264-f7962180-0b97-11ea-97b3-8c57fc2dcd84.png)
![image](https://user-images.githubusercontent.com/17438227/69230544-886cfd00-0b98-11ea-8a11-53f1c8669b64.png)
- для валидации ссылки лучше использовать пакет `validator` и его метод isURL

Хорошо:
- Сделано дополнительное задание.
- в репозитории понятный и незапутанный gitflow.

Graev committed:
Correct the error

Исправлены ошибки линтинга;
Исправленна некорректная работа API;
При создании ресурсов статус ответа теперь 201 (был 200);
Добавлена прослойка в роутинге (сейчас не сильно заметна разница, но при маштабировании будет удобнее);
Коллбеки в контроллерах заменены на промиссы (если честно, не нашел нормальной документации по orFail);
Для валидации ссылок используется метод isURL пакета validator.

Boortcore commented:
Необходимо исправить:
- API возвращает 200 при попытке получить юзера по валидному ObjectId "5dd510024c38a003fc91216b", которого не существует в базе
Должен быть статус 404
![image](https://user-images.githubusercontent.com/17438227/69267157-a6a91c00-0bdd-11ea-9902-93335e60b656.png)
- API возвращает 200 при попытке удалить карточку по валидному ObjectId "5dd2a79e32bef703703db415", которого не существует в базе |
![image](https://user-images.githubusercontent.com/17438227/69267209-c0e2fa00-0bdd-11ea-8da0-10e128bf5941.png)
- при попытке удалить карточку по невалидному id сервер выдаёт предупреждение: " UnhandledPromiseRejectionWarning: Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client" (как и при попытке получить юзера по валидному ObjectId "5dd510024c38a003fc91216b", которого не существует в базе

Ошибки eslint исправлены.

Graev committed:
Correct the new errors

Добавил статус при валидном, но не правильном id;
Добавил отлов невалидных id

Graev commented:
Ошибку UnhandledPromiseRejectionWarning: Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client" у меня получить не удалось(может быть не там смотрю), надеюсь, что после добавления отлова невалидного id она исправилась.

Boortcore commented:
Необходимо исправить:
Обработку поиска юзера по id, которого нет в базе, можно сделать следующим образом:
![image](https://user-images.githubusercontent.com/17438227/69406106-4c5da780-0d12-11ea-8f5c-53ac82a569c6.png)
Аналогично нужно сделать и с функцией удаления карточек. 
У orFail несколько иное предназначение. Этот метод нужен для того, чтобы сгенерировать ошибку, которая должна обработаться в блоке catch, а отправка ответа на клиент осуществляется в блоке catch.
![image](https://user-images.githubusercontent.com/17438227/69407098-c2630e00-0d14-11ea-9f4f-dd0d931dff09.png)

Текущая реализация пробрасывает из onFail в catch ошибку, у которой нет свойства message. Сервер из-за этого выбрасывал необрабатываемое исключение, т.к. вызов err.message.includes, где `err.message === undefined` приведёт к ошибке типа

Graev committed:
Changed bug catching

Заменил некорректное использование orFail, на проверяющий if в then

Graev commented:
Все, теперь понял предназначение и работу orFail. Спасибо большое!
